File: TubeFromCenterline/utility/vtk_binary_to_ascii.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
バイナリ形式の legacy VTK (*.vtk) ファイルを読み込み、
ASCII 形式の legacy VTK (*.vtk) として保存するスクリプト（GUI版）。

- 起動するとファイル選択ダイアログが開くので、入力 .vtk を選択する。
- 続いて、出力ファイル名を指定するダイアログが開く。
"""

import os
import sys

# --- VTK 読み込み ---
try:
    import vtk
except ImportError:
    print("vtk モジュールが見つかりません。先に 'pip install vtk' を実行してください。")
    sys.exit(1)

# --- Tkinter でファイルダイアログ ---
import tkinter as tk
from tkinter import filedialog, messagebox


def convert_binary_vtk_to_ascii(input_path: str, output_path: str) -> None:
    # 入力ファイル存在チェック
    if not os.path.isfile(input_path):
        raise FileNotFoundError(f"入力ファイルが見つかりません: {input_path}")

    # legacy VTK 用のリーダー
    reader = vtk.vtkDataSetReader()
    reader.SetFileName(input_path)
    reader.ReadAllScalarsOn()
    reader.ReadAllVectorsOn()
    reader.ReadAllTensorsOn()
    reader.ReadAllFieldsOn()

    print(f"読み込み中: {input_path}")
    reader.Update()

    data = reader.GetOutput()
    if data is None:
        raise RuntimeError("VTK ファイルからデータを取得できませんでした。legacy VTK 形式か確認してください。")

    # DataSetWriter を使って ASCII 形式で書き出し
    writer = vtk.vtkDataSetWriter()
    writer.SetFileName(output_path)
    writer.SetInputData(data)
    writer.SetFileTypeToASCII()  # ← ここで ASCII 指定

    print(f"書き出し中 (ASCII): {output_path}")
    if writer.Write() == 0:
        raise RuntimeError("VTK ファイルの書き出しに失敗しました。")

    print("変換完了。")


def main():
    # Tk のルートウィンドウ（非表示）
    root = tk.Tk()
    root.withdraw()

    # --- 入力ファイル選択 ---
    input_path = filedialog.askopenfilename(
        title="バイナリ形式の VTK ファイルを選択してください",
        filetypes=[("VTK files", "*.vtk"), ("All files", "*.*")]
    )

    if not input_path:
        # キャンセル
        print("入力ファイルが選択されませんでした。終了します。")
        return

    # デフォルト出力ファイル名を作る
    base, ext = os.path.splitext(input_path)
    default_output = base + "_ascii.vtk"

    # --- 出力ファイル名選択 ---
    output_path = filedialog.asksaveasfilename(
        title="ASCII 形式 VTK の保存先を選択してください",
        initialfile=os.path.basename(default_output),
        defaultextension=".vtk",
        filetypes=[("VTK files", "*.vtk"), ("All files", "*.*")]
    )

    if not output_path:
        print("出力ファイルが指定されませんでした。終了します。")
        return

    try:
        convert_binary_vtk_to_ascii(input_path, output_path)
    except Exception as e:
        messagebox.showerror("エラー", f"変換中にエラーが発生しました。\n{e}")
        print("エラー:", e)
        return

    messagebox.showinfo("完了", "変換が正常に完了しました。")
    print("変換が正常に完了しました。")


if __name__ == "__main__":
    main()

File: TubeFromCenterline/utility/VtkBinaryToAscii.cpp
#include <vtkDataSet.h>
#include <vtkDataSetReader.h>
#include <vtkDataSetWriter.h>
#include <vtkNew.h>

#include <iostream>
#include <string>

#if defined(_WIN32)
// Windows のファイルダイアログ用
#include <windows.h>
#include <commdlg.h>
#include <filesystem>
#include <cstdio>

// 入力 VTK ファイルを選ぶダイアログ
static std::string OpenVTKFileDialog()
{
    char fileBuf[MAX_PATH] = "";
    OPENFILENAMEA ofn{};
    ofn.lStructSize = sizeof(ofn);
    ofn.hwndOwner   = nullptr;
    ofn.lpstrFilter = "VTK Files (*.vtk)\0*.vtk\0All Files (*.*)\0*.*\0";
    ofn.lpstrFile   = fileBuf;
    ofn.nMaxFile    = MAX_PATH;
    ofn.Flags       = OFN_FILEMUSTEXIST | OFN_PATHMUSTEXIST;

    if (GetOpenFileNameA(&ofn))
    {
        return std::string(fileBuf);
    }
    return std::string();
}

// 出力 VTK ファイルを選ぶダイアログ
static std::string SaveVTKFileDialog(const std::string& defaultName)
{
    char fileBuf[MAX_PATH] = "";
    std::snprintf(fileBuf, MAX_PATH, "%s", defaultName.c_str());

    OPENFILENAMEA ofn{};
    ofn.lStructSize = sizeof(ofn);
    ofn.hwndOwner   = nullptr;
    ofn.lpstrFilter = "VTK Files (*.vtk)\0*.vtk\0All Files (*.*)\0*.*\0";
    ofn.lpstrFile   = fileBuf;
    ofn.nMaxFile    = MAX_PATH;
    ofn.Flags       = OFN_OVERWRITEPROMPT;
    ofn.lpstrDefExt = "vtk"; // 拡張子補完

    if (GetSaveFileNameA(&ofn))
    {
        return std::string(fileBuf);
    }
    return std::string();
}
#endif // _WIN32

// 実際の変換処理
static bool ConvertBinaryVTKToASCII(const std::string& inputFile,
                                    const std::string& outputFile)
{
    vtkNew<vtkDataSetReader> reader;
    reader->SetFileName(inputFile.c_str());
    reader->ReadAllScalarsOn();
    reader->ReadAllVectorsOn();
    reader->ReadAllTensorsOn();
    reader->ReadAllFieldsOn();

    std::cout << "Reading: " << inputFile << std::endl;
    reader->Update();

    vtkDataSet* data = reader->GetOutput();
    if (!data)
    {
        std::cerr << "Failed to read dataset from: " << inputFile << std::endl;
        return false;
    }

    vtkNew<vtkDataSetWriter> writer;
    writer->SetFileName(outputFile.c_str());
    writer->SetInputData(data);
    writer->SetFileTypeToASCII();  // ← ASCII 形式で出力

    std::cout << "Writing ASCII VTK: " << outputFile << std::endl;
    if (!writer->Write())
    {
        std::cerr << "Failed to write ASCII VTK to: " << outputFile << std::endl;
        return false;
    }

    std::cout << "Conversion finished successfully." << std::endl;
    return true;
}

int main(int argc, char* argv[])
{
    std::string inputFile;
    std::string outputFile;

#if defined(_WIN32)
    // Windows:
    //   - 引数が 2つ (input, output) あればそれを使う
    //   - 引数が足りなければ GUI ダイアログで選択
    if (argc < 3)
    {
        inputFile = OpenVTKFileDialog();
        if (inputFile.empty())
        {
            std::cerr << "Input file was not selected. Exiting." << std::endl;
            return EXIT_FAILURE;
        }

        // デフォルト出力ファイル名: inputFile の末尾に "_ascii.vtk" を付ける
        std::filesystem::path inPath(inputFile);
        std::string defaultOut =
            (inPath.parent_path() / (inPath.stem().string() + "_ascii.vtk")).string();

        outputFile = SaveVTKFileDialog(defaultOut);
        if (outputFile.empty())
        {
            std::cerr << "Output file was not selected. Exiting." << std::endl;
            return EXIT_FAILURE;
        }
    }
    else
    {
        inputFile  = argv[1];
        outputFile = argv[2];
    }
#else
    // macOS / Linux: コマンドライン引数で指定
    if (argc < 3)
    {
        std::cerr << "Usage: " << argv[0]
                  << " input_binary.vtk output_ascii.vtk" << std::endl;
        return EXIT_FAILURE;
    }
    inputFile  = argv[1];
    outputFile = argv[2];
#endif

    if (!ConvertBinaryVTKToASCII(inputFile, outputFile))
    {
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
